classes/TaskRunner/Upgrade/UpgradeFiles.php

```diff
@@ -0,0 +1,284 @@
+                    return false;
+                }
+            } else { // directory already exists
+                $this->logger->debug($this->translator->trans('Directory %s already exists.', array($file), 'Modules.Autoupgrade.Admin'));
+                return true;
+            }
+        } elseif (is_file($orig)) {
+            $translationAdapter = $this->container->getTranslationAdapter();
+            if ($translationAdapter->isTranslationFile($file) && file_exists($dest)) {
+                $type_trad = $translationAdapter->getTranslationFileType($file);
+                if ($translationAdapter->mergeTranslationFile($orig, $dest, $type_trad)) {
+                    $this->logger->info($this->translator->trans('[TRANSLATION] The translation files have been merged into file %s.', array($dest), 'Modules.Autoupgrade.Admin'));
+                    return true;
+                }
+                $this->logger->warning($this->translator->trans(
+                    '[TRANSLATION] The translation files have not been merged into file %filename%. Switch to copy %filename%.',
+                    array('%filename%' => $dest),
+                    'Modules.Autoupgrade.Admin'
+                ));
+            }
+
+            // upgrade exception were above. This part now process all files that have to be upgraded (means to modify or to remove)
+            // delete before updating (and this will also remove deprecated files)
+            if (copy($orig, $dest)) {
+                $this->logger->debug($this->translator->trans('Copied %1$s.', array($file), 'Modules.Autoupgrade.Admin'));
+                return true;
+            } else {
+                $this->next = 'error';
+                $this->logger->error($this->translator->trans('Error while copying file %s', array($file), 'Modules.Autoupgrade.Admin'));
+                return false;
+            }
+        } elseif (is_file($dest)) {
+            if (file_exists($dest)) {
+                unlink($dest);
+            }
+            $this->logger->debug(sprintf('removed file %1$s.', $file));
+            return true;
+        } elseif (is_dir($dest)) {
+            if (strpos($dest, DIRECTORY_SEPARATOR.'modules'.DIRECTORY_SEPARATOR) === false) {
+                FilesystemAdapter::deleteDirectory($dest, true);
+            }
+            $this->logger->debug(sprintf('removed dir %1$s.', $file));
+            return true;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * First call of this task needs a warmup, where we load the files list to be upgraded.
+     * 
+     * @return boolean
+     */
+    protected function warmUp()
+    {
+        $admin_dir = str_replace($this->container->getProperty(UpgradeContainer::PS_ROOT_PATH).DIRECTORY_SEPARATOR, '', $this->container->getProperty(UpgradeContainer::PS_ADMIN_PATH));
```

Comment:
You can maybe set a variable containing `$this->container->getProperty(UpgradeContainer::LATEST_PATH).DIRECTORY_SEPARATOR` to make this page more readable.